Property wrapper using a keypath leads to "Struct declaration cannot close over value '$kp$' defined in outer scope"

**Describe the bug**
In current nightly Swift toolchains, the use of certain keypaths with property wrappers now causes a compilation error of "Struct declaration cannot close over value '$kp$' defined in outer scope". I believe this is a fairly recent regression, with the 2022-08-09 nightly snapshot working as expected but the 2022-09-24 starting to experience this (sorry, don't have finer granularity than that).

The following single-file will trigger this compilation error:

```swift
@propertyWrapper
struct Option {
    let help: String
    var value: Double = 0.0

    var wrappedValue: Double {
        get { value }
        set { value = newValue }
    }
}

enum TestEnum: String, CaseIterable {
    case hello = "hello"
    case world = "world"
}

struct TestStruct {
    init() {}

    @Option(help: "Values: \(TestEnum.allCases.map(\.rawValue))") // <- "Struct declaration cannot close over value '$kp$' defined in outer scope"
    var value
}
```

**Steps To Reproduce**
Steps to reproduce the behavior:
1. Place the above code in a file.
2. Build it using a nightly Swift toolchain newer than at least 2022-09-24.

**Expected behavior**
This code should build cleanly, as it did in prior nightly snapshots.

**Environment (please fill out the following information)**
 - OS: macOS 12.5.1
 - Swift nightly toolchains at least from 2022-09-24 to 2022-10-09

@xedin It looks like my changes expose a latent general issue in property wrapper decl checking. Capture lists raise spurious errors in general when they appear in property wrapper initializers:

```
@propertyWrapper
struct ButtWrapper {
    init(_: String) {}

    var wrappedValue: Int { return 0 }
}

struct Butt {
    @ButtWrapper("\([1,2,3].map({[x = 42] in "\($0 + x)" }).reduce("", +))")
    var foo: Int
}
```

raises the error:

```
/Users/jgroff/bar.swift:9:54: error: struct declaration cannot close over value 'x' defined in outer scope
    @ButtWrapper("\([1,2,3].map({[x = 42] in "\($0 + x)" }).reduce("", +))")
                                                     ^
```

--- a/lib/Sema/TypeCheckStmt.cpp
+++ b/lib/Sema/TypeCheckStmt.cpp
@@ -98,9 +98,13 @@ namespace {
         return Action::SkipChildren(E);
       } 
 
-      // Capture lists need to be reparented to enclosing autoclosures.
       if (auto CapE = dyn_cast<CaptureListExpr>(E)) {
-        if (isa<AutoClosureExpr>(ParentDC)) {
+        // Capture lists need to be reparented to enclosing autoclosures
+        // and/or initializers of property wrapper backing properties
+        // (because they subsume initializers associated with wrapped
+        // properties).
+        if (isa<AutoClosureExpr>(ParentDC) ||
+            isPropertyWrapperBackingPropertyInitContext(ParentDC)) {
           for (auto &Cap : CapE->getCaptureList()) {
             Cap.PBD->setDeclContext(ParentDC);
             Cap.getVar()->setDeclContext(ParentDC);
@@ -135,6 +139,21 @@ namespace {
       // But we do want to walk into the initializers of local
       // variables.
       return Action::VisitChildrenIf(isa<PatternBindingDecl>(D));
+    }
+
+  private:
+    static bool isPropertyWrapperBackingPropertyInitContext(DeclContext *DC) {
+      auto *init = dyn_cast<PatternBindingInitializer>(DC);
+      if (!init)
+        return false;
+
+      if (auto *PB = init->getBinding()) {
+        auto *var = PB->getSingleVar();
+        return var && var->getOriginalWrappedProperty(
+                          PropertyWrapperSynthesizedPropertyKind::Backing);
+      }
+
+      return false;
     }
   };
 

--- a/test/Sema/property_wrappers.swift
+++ b/test/Sema/property_wrappers.swift
@@ -83,3 +83,40 @@ struct S2_58201 {
   // CHECK:            autoclosure_expr implicit type='() -> Bool' location={{.*}}.swift:[[@LINE+1]]:26 range=[{{.+}}] discriminator=1 escaping
   @W_58201 var b: Bool = false
 }
+
+// https://github.com/apple/swift/issues/61570
+// rdar://problem/101813792 - Check that captured variables are re-parented to backing var initializer
+do {
+  @propertyWrapper
+  struct Wrapper {
+    init(_: String) {}
+
+    var wrappedValue: Int { return 0 }
+  }
+
+  struct TestExplicitCapture {
+    @Wrapper("\([1,2,3].map({[x = 42] in "\($0 + x)" }).reduce("", +))")
+    var wrapped: Int // Ok, becomes var _wrapped: Wrapper<Int> = Wrapper(wrappedValue: "\([1,2,3].map({[x = 42] in "\($0 + x)" }).reduce("", +)))
+  }
+
+  @propertyWrapper
+  struct Option {
+    let help: String
+    var value: Double = 0.0
+
+    var wrappedValue: Double {
+      get { value }
+      set { value = newValue }
+    }
+  }
+
+  enum TestEnum: String, CaseIterable {
+    case hello = "hello"
+    case world = "world"
+  }
+
+  struct TestImplicitCapture {
+    @Option(help: "Values: \(TestEnum.allCases.map(\.rawValue))")
+    var value // Ok - $kp$ of key path is captured implicitly by backing variable init
+  }
+}

